### PR TITLE
Add v2 notifications routes

### PR DESCRIPTION
**Pull request by iamacook**
## Summary

This adds SiWe-guard routes for upserting push notification subscriptions/getting preferences, as well as deletion routes for Safe/device subscriptions.

### `POST` `/v2/register/notifications` (guarded)

The `cloudMessagingToken` is the Firebase Cloud Messaging token for the device being registered. A UUID can be provided for this device, or one will be returned.

Subscriptions are provided per chain-specific Safe, alongside the preferences of which notification types should be subscribed-to.

Note: the given preferences overwrites existing ones if upserting a subscription.

```ts
enum NotificationType {
  CONFIRMATION_REQUEST = 'CONFIRMATION_REQUEST',
  DELETED_MULTISIG_TRANSACTION = 'DELETED_MULTISIG_TRANSACTION',
  EXECUTED_MULTISIG_TRANSACTION = 'EXECUTED_MULTISIG_TRANSACTION',
  INCOMING_ETHER = 'INCOMING_ETHER',
  INCOMING_TOKEN = 'INCOMING_TOKEN',
  MESSAGE_CONFIRMATION_REQUEST = 'MESSAGE_CONFIRMATION_REQUEST',
  MODULE_TRANSACTION = 'MODULE_TRANSACTION',
}

type Request = {
  cloudMessagingToken: string;
  safes: Array<{
    chainId: string;
    address: `0x${string}`;
    notificationTypes: Array<NotificationType>;
  }>
  deviceType: 'ANDROID' | 'IOS' | 'WEB';
  deviceUuid?: `${string}-${string}-${string}-${string}-${string}`;
}

type Response = {
  deviceUuid: `${string}-${string}-${string}-${string}-${string}`;
}
```

### `GET` `/v2/chains/:chainId/notifications/devices/:deviceUuid/safes/:safeAddress` (guarded)

This returns an array of `NotificationTypes` for the respective Safe.

```ts
type Response = Array<NotificationType>
```

### `DELETE` `/v2/chains/:chainId/notifications/devices/:deviceUuid/safes/:safeAddress`

This deletes the device-specific subscription for the given Safe.

### `DELETE` `/v2/chains/:chainId/notifications/devices/:deviceUuid`

This deleted the device, and therefore all subscriptions associated with it.

## Changes

- Create above routes in a new `NotificationsModuleV2` with associated service and controller, tweaking the respective repository/datasource in accordance.
- Add relative validation for entities:
  - Create `UpsertSubscriptionsDto`, located on the route level and infer the type from it.
  - Create `UuidSchema`.
- Remove `Uuid` entity in place of `crypto.UUID`.
- Add appropriate tests.

### CODE REVIEW

File: src/domain/notifications/notifications.repository.v2.ts [hunk 1]
```diff
@@ -63,7 +66,13 @@ export class NotificationsRepositoryV2 implements INotificationsRepositoryV2 {
       );
     } catch (e) {
       if (this.isTokenUnregistered(e)) {
-        await this.deleteDevice(args.deviceUuid).catch(() => null);
+        this.loggingService.info(
+          `Deleting unregistered token for device ${args.deviceUuid}: ${e}`,
+        );
```
Comment (hectorgomezv) [inline src/domain/notifications/notifications.repository.v2.ts:71]: Maybe I'm not getting it right, but this message is stating we are deleting an unregistered token, and actually the device is being deleted, isn't it? (according to the subsequent datasource function call)

File: src/domain/notifications/notifications.repository.v2.ts [hunk 2]
```diff
@@ -104,23 +129,29 @@ export class NotificationsRepositoryV2 implements INotificationsRepositoryV2 {
...
     safeAddress: `0x${string}`;
-    signerAddress: `0x${string}`;
   }): Promise<void> {
-    return this.notificationsDatasource.deleteSubscription(args);
+    return this.notificationsDatasource.deleteSubscription({
+      deviceUuid: args.deviceUuid,
+      chainId: args.chainId,
+      safeAddress: args.safeAddress,
+    });
   }
```
Comment (hectorgomezv) [inline src/domain/notifications/notifications.repository.v2.ts:152]: Nit: we could just still `return this.notificationsDatasource.deleteSubscription(args)` here, right?

File: src/datasources/notifications/notifications.datasource.ts [hunk 3]
```diff
@@ -242,14 +236,14 @@ export class NotificationsDatasource implements INotificationsDatasource {
...
-  async deleteDevice(deviceUuid: Uuid): Promise<void> {
+  async deleteDevice(deviceUuid: UUID): Promise<void> {
     await this.sql`
       DELETE FROM push_notification_devices
       WHERE device_uuid = ${deviceUuid}
     `.catch((e) => {
       const error = 'Error deleting device';
       this.loggingService.info(`${error}: ${asError(e).message}`);
-      throw new UnprocessableEntityException(error);
+      throw new NotFoundException(error);
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/datasources/notifications/notifications.datasource.ts]: Good point. I reverted this in 2a1216101fd50bafd404c24e2a6fddae907dcbf4.

File: src/datasources/notifications/notifications.datasource.ts [hunk 4]
```diff
@@ -242,14 +236,14 @@ export class NotificationsDatasource implements INotificationsDatasource {
...
    * @param deviceUuid Device UUID
    */
-  async deleteDevice(deviceUuid: Uuid): Promise<void> {
+  async deleteDevice(deviceUuid: UUID): Promise<void> {
     await this.sql`
       DELETE FROM push_notification_devices
       WHERE device_uuid = ${deviceUuid}
     `.catch((e) => {
       const error = 'Error deleting device';
       this.loggingService.info(`${error}: ${asError(e).message}`);
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/datasources/notifications/notifications.datasource.ts]: Changed in 2a1216101fd50bafd404c24e2a6fddae907dcbf4.

File: src/domain/notifications/notifications.repository.v2.ts (hunk 1 above)
Comment (iamacook) [reply src/domain/notifications/notifications.repository.v2.ts]: I adjusted the message in 2a1216101fd50bafd404c24e2a6fddae907dcbf4 to explain that we're deleting a device _because of_ the token being stale.

File: src/domain/notifications/notifications.repository.v2.ts (hunk 2 above)
Comment (iamacook) [reply src/domain/notifications/notifications.repository.v2.ts]: Changed in 2a1216101fd50bafd404c24e2a6fddae907dcbf4.

**Review by hectorgomezv — APPROVED** [review]

**Review by hectorgomezv — APPROVED** [review]

**Pull request merged**